src/legacy/index.ts

```diff
@@ -115,19 +123,24 @@ async function buildGraph(
       name: pkgName,
       version: depTree.version,
     };
-    pkgNodeId = getNodeId(pkg.name, pkg.version, treeHash);
+    pkgNodeId = hashBasedNodeId(pkg.name, pkg.version, treeHash);
 
     const nodeInfo: types.NodeInfo = {};
 
     if (depTree.versionProvenance) {
       nodeInfo.versionProvenance = depTree.versionProvenance;
     }
 
-    builder.addPkgNode(pkg, pkgNodeId, nodeInfo);
+    // If a node with a matching hash-based ID already exists in the graph,
+    // we have already processed an identical subtree and there is no need to add the edges
+    // for the node's dependencies.
+    shouldAddEdges = builder.addPkgNode(pkg, pkgNodeId, nodeInfo);
   }
 
-  for (const depNodeId of depNodesIds) {
-    builder.connectDep(pkgNodeId, depNodeId);
+  if (shouldAddEdges) {
```

Comment:
I'm not sure how confident I feel about this... is the preexistence of a node really indicative of whether or not it has been connected?

Reply:
docs for `graphlib::setEdge` state `Complexity: O(1)`.. I wonder how much time this optimisation saves?

Reply:
20% roughly.

"is the preexistence of a node really indicative of whether or not it has been connected?" - yes, because this transformation is DFS, so if a node with a hash exists, then it has been already created and properly connected.

Reply:
> yes, because this transformation is DFS
yep, right now, but what I'm getting at is that we're relying on an implementation detail here. Perhaps I'm being overly paranoid

Reply:
I guess all sorts of tests will fail if we break this in the future.. so ✅